workflow/session.go

```diff
@@ -0,0 +1,112 @@
+//         support. By default, 1000 is used.
+
+// CreateSession creates a session and return a new context which contains information
+// of the created session. The session will be created on the tasklist user specified in
+// ActivityOptions. If none is specified, the default one will be used.
+//
+// CreationSession will fail in the following situations:
+//     1. The context passed in already contains a session which is still open
+//        (not closed and failed).
+//     2. The number of sessions currently running on the worker has reached
+//        MaxConCurrentSessionExecutionSize, which is specified when starting the worker.
+//
+// If an activity is executed using the returned context, it's regarded as part of the
+// session and all activities within a session will be executed by the same worker.
+// User still needs to handle the error returned when executing an activity. Session will
+// not be marked as failed if an activity within it returns an error. Only when the worker
+// executing the session is down, that session will be marked as failed. Executing an activity
+// within a failed session will return an [TODO] error immediately without even scheduling the activity.
+//
+// If user wants to end a session since some activity returns an error, use the CompleteSession API below,
+// and new session can be created if necessary to retry the whole session.
+//
+// Example:
+//    sessionCtx, err := CreateSession(ctx)
+//    if err != nil {
+//		    // Creation failed. Wrong ctx or too many outstanding sessions.
+//    }
+//    err = ExecuteActivity(sessionCtx, someActivityFunc, activityInput).Get(sessionCtx, nil)
+//    if err != nil {
+//        // Session has failed or activity itself failed.
+//    }
+//    ... // execute more activities using sessionCtx
+//    err = CompleteSession(sessionCtx)
+//    if err != nil {
+//        // Wrong ctx is used or failed to release session resource.
+//    }
+func CreateSession(ctx Context) (Context, error) {
+	return internal.CreateSession(ctx)
+}
+
+// RecreateSession recreate a session based on the sessionInfo passed in. Activities executed within
+// the recreated session will be executed by the same worker as the previous session. RecreateSession()
+// returns an error under the same situation as CreateSession() and has the same usage as CreateSession().
+// It will not check the state of the session described by the sessionInfo passed in, so user can recreate
+// a session based on a failed or completed session.
+//
+// The main usage of RecreateSession is for long sessions that are splited into multiple runs. At the end of
+// one run, complete the current session, get sessionInfo from the context and pass the information to the
+// next run. In the new run, the session can be recreated on the information.
+func RecreateSession(ctx Context, sessionInfo *SessionInfo) (Context, error) {
```

Comment:
Let's rename to CreateSessionForResourceID.

Reply:
Since we need to hide resourceID from user for now, is it ok to keep this name?